Add boolean expression parser

SpaceKatt opened the pull request:
Creates a truth table from a boolean expression

SpaceKatt committed:
Add temp vim files to ignore

SpaceKatt committed:
Add structure for future functionality

SpaceKatt committed:
Implement boolean expression parser

Invalid expressions break parser. Refactoring of precedence decisions
is in order.

SpaceKatt committed:
Fix invalid expressions

It APPEARS to be working, more testing is needed...

SpaceKatt committed:
Add check for valid variable placement, update formatting

Variables can no longer be placed directly next to one another

SpaceKatt committed:
Simplify some minor stuff; Backup before making changes

!(x + y) is not giving the right result. For some reason,
it is parsing to be xy+ in post-fix. Going to change this next

SpaceKatt committed:
Integrate command into Discord API

Tested on a Discord server, everything seems to work!

SpaceKatt committed:
Update docs

SpaceKatt committed:
Merge branch 'boolean'

Add __bool command to discord bot

Chris-Johnston commented on .gitignore:
```diff
@@ -164,3 +164,6 @@ crashlytics.properties
 crashlytics-build.properties
 fabric.properties
 
+# Vim temp files
```
I just found this, might be a more comprehensive .gitignore

https://github.com/github/gitignore/blob/master/Global/Vim.gitignore

SpaceKatt replied:
Most likely, I've never run into a situation where I need more exception, but it would be good to include

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+
+    # Defines operator precedence
+    prec_dict = {
+        '!' : 0, # NOT
```
Can we alias `'~'` as NOT as well

SpaceKatt replied:
Yes, that could be added in pretty easily

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+        self.bot = bot
+
+    @commands.command()
+    async def bool(self, ctx, exp):
```
Actually, if you change this to `async def bool(self, ctx, *, exp):` you don't need to wrap in quotes. It's in the discord API docs, but also can be seen here: https://github.com/Chris-Johnston/CSSBot_Py/blob/master/cogs/courseInfo.py#L59

SpaceKatt replied:
Huh, that is pretty neat. I'll change that

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+#####################################################################
+#####################################################################
+
+class BooleanExpr:
```
Should we move this to another file? ~~(Could also make easier to test down the line if/when that's implemented)~~ I just saw that you have tests, nice!

SpaceKatt replied:
I originally had it in another file, but I'm sick and didn't feel like dealing with creating a module structure =p

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+    # Defines operator precedence
+    prec_dict = {
+        '!' : 0, # NOT
+        '*' : 1, # AND
```
Does this interpret `AB` as `A*B`?

SpaceKatt replied:
No. I don't think it would be worth the effort to implement that behavior.

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+                self.error = True
+                self.error_msg = "{} is not a valid symbol!".format(char)
+        if self.error == True:
+            return
```
Would it make more sense to throw an exception here instead?

SpaceKatt replied:
I'm unaware of how the Discord API handles exceptions and I wanted an easy way to display what was wrong with the expression. Throwing an exception would work well, I just don't know how to display that information back to a user in Discord

Chris-Johnston replied:
iirc, the discord.py library being used catches all exceptions for you, and there's a way provided by the framework for logging or sending messages on exceptions that I'm not using.

What might be simpler is to wrap the call to this command (in the bot-specific code) with a try catch, and reply back if there was some parsing error with invalid input.

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+
+    def is_valid_op(self, char):
+        '''True if char is a valid boolean operator'''
+        result = char == '+' or char == '^' or char == '*' or char == '!'
```
Should we make these class constants as a list and then do something like `return char in boolean_operators`

SpaceKatt replied:
That would be a much more pythonic way to handle that

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+            pass
+        # Else, pop two var, add one var to stack
+        else:
+            if len(self.var_stack) < 2:
```
Simplify to `if temp != '!':` ?

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+                if char == '^':
+                    one = post_stack.pop()
+                    two = post_stack.pop()
+                    result = one and not two or two and not one
```
You *should* be able to do `bool(one) ^ bool(two)`

https://stackoverflow.com/questions/432842/how-do-you-get-the-logical-xor-of-two-variables-in-python

SpaceKatt replied:
Sure, why not =p

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+        self.result_formatted += " " * len(self.disp_exp) + "|\n"
+
+    def bool_to_int(self, booly):
+        '''Returns 1 if True, 0 if False'''
```
```py
>>> int(True)
1
>>> int(False)
0
```

Let's use int() instead.

SpaceKatt replied:
Gotchya, I didn't know that worked like that

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+        return self.result_formatted
+
+# Run some basic tests
+if __name__ == '__main__':
```
Awesome, but maybe we should use a testing framework for this?

SpaceKatt replied:
Probably, I'll have to modify the class to provide some functions to make for non-gross testing. I'm not about to write tests to match formatted strings, so probably will create a `get_bool_results()` method or something to facilitate testing

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+                self.error_msg = "'!' should not follow a ')'"
+                return False
+            if self.orig_exp[i] == '!' and self.orig_exp[i - 1] == '!':
+                self.error_msg = "Please remove redundant ! operators..."
```
What about doing `str.Replace('!!', '')`, as a easy way to catch stuff like this

SpaceKatt replied:
That sounds like a dandy idea

Chris-Johnston commented on cogs/boolean_utils.py:
```diff
@@ -0,0 +1,395 @@
+        '''Returns formatted version of the truth table for a boolean expr'''
+        if not self.valid_vars():
+            self.error = True
+            self.error_msg = "Cannot have two variables in a row..."
```
Would it be possible to interpret this as `A*B`?

SpaceKatt replied:
Probably, but I don't think it would be worth the effort imo

Chris-Johnston replied:
Probably could open that up as something to do "down the line" after this is merged

assuming anyone of us has time to work on it.

Chris-Johnston reviewed (changes requested):
Looks great, will test on my machine as well. I was picky over a few minor details.

SpaceKatt committed:
Update .gitignore for vimfiles

SpaceKatt committed:
Add `~` as alias for NOT.

Also fix function signature so one does not have to wrap functions
in quotes in discord.

SpaceKatt committed:
Make simplifications in the logic

SpaceKatt committed:
Merge branch 'master' into master

SpaceKatt commented:
To do: 

- Add testing framework
- Separate classes into different files
- Change error handling to exceptions
- Implement AB as A * B

Chris-Johnston commented:
Let's ship it. I'll leave it to you to get back to the todo list. (Open an issue?)

Chris-Johnston reviewed (commented):
how do you even approve changes in github
